In [9]:
+
 """Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio
test. The function should return the F-value and the corresponding p-value for the test.
"""

# Solution:-

import numpy as np
import scipy.stats as stat

def f_var_test(data1,data2):
    
    n1 = len(data1)
    n2 = len(data2)
    
    var1 = np.var(data1, ddof =1)
    var2 = np.var(data2, ddof =1)
    
    if var1>var2:
        f_stat = var1/var2
        dof1 = n1-1
        dof2 = n2-1
    else:
        f_stat = var2/var1
        dof1 = n2-1
        dof2 = n1-1
    
    p_val = stat.f.sf(f_stat, dof1, dof2)
    
    return f_stat, p_val
        

In [10]:
data1 = [1, 2, 3, 4, 5]
data2 = [3, 5, 7, 9, 11]
f_stat, p_val = f_var_test(data1, data2)
print("F-value:", f_stat)
print("p-value:", p_val)

F-value: 4.0
p-value: 0.10400000000000002


-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [12]:
"""Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an
F-distribution, write a Python function that returns the critical F-value for a two-tailed test.
"""

import scipy.stats as stat



def F_value(alpha, dof1, dof2):
    return stat.f.ppf(alpha/2, dof1, dof2), stat.f.ppf(1-alpha/2, dof1, dof2)    # returns left and right-extreme  critical value

In [15]:
alpha = 0.5

# assuming dof:

dof1 = 2
dof2 = 5

lower_cv, upper_cv = F_value(alpha, dof1, dof2)

print(f"Critical Value: ({lower_cv}, {upper_cv})")

Critical Value: (0.3048878636154987, 1.8527528164806204)


----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [4]:
"""Q3. Write a Python program that generates random samples from two normal distributions with known

variances and uses an F-test to determine if the variances are equal. The program should output the F-
value, degrees of freedom, and p-value for the test.
"""

# Solution:-

import numpy as np
import scipy.stats as stat

# let known variances be:-

sigma1 = 1
sigma2 = 2

mu1 = 0
mu2 = 1

n1 = 30
n2 = 40

# creating random samples:-

np.random.seed(43)   # setting seed value to reproduce sample value for experiment

sample1 = np.random.normal(mu1, sigma1, n1)
sample2 = np.random.normal(mu2, sigma2, n1)

In [8]:
f = np.var(sample1, ddof = 1)/np.var(sample2, ddof = 1)

df1 = n1 -1
df2 = n2 -1

p_val = 1 - stat.f.cdf(f, df1, df2)


print(f"F-Value: {f}")
print(f"\ndegress of freedom: df1: {df1}, df2: {df2}")
print(f"\nP-Value: {p_val}")

F-Value: 0.17896547355865366

degress of freedom: df1: 29, df2: 39

P-Value: 0.9999964238077438


-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

Ans. 

Null Hypothesis(H0): var1 = var2

Alternate Hypothesis(H1): var1 != var2

Performing F-stats we get:

F = 10/15

= 0.667

For significancle level of 5% and dof1 = dof2 => 12-1 => 11  we can calculate critical value to be:-

critical value = 2.83

Conclusion:-

Since, F-val < critical value, we fail to reject Null Hypothesis and can say that we don't have enough evidence to suggest that there is significant difference between variance of two population.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

Ans. 

Null Hypothesis(H0):-   var = 0.005

Alternate Hypothesis(H1):-  Variance is greater than 0.005

Performing F-test we get:-

F-Value = sample variance/population variance

= 0.006/0.005

= 1.2

From given question we can see:-

dof sample = 25-1 => 24

dof population = infinite

For significance levele of alpha = 0.01

critical value = 1.791

Conclusion:-

Since F-val(1.2) < Critical Value(1.791) we cannot reject Null Hypothesis and we can say that we don't have enough evidence to say that population variance is greater than 0.005 for significance level of 1%.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [9]:
"""Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an
F-distribution and calculates the mean and variance of the distribution. The function should return the
mean and variance as a tuple.
"""

def fdist_mean_var(dofn, dofd):
    
    if dofn <= 0 or dofd <= 0:
        raise ValueError("degree of freedom must be greater than 0")
    
    mean = dofd/(dofd-2)
    
    variance =  (2 * (dofd**2) * (dofn + dofd - 2)) / ((dofn * (dofd - 2)**2 * (dofd - 4)))
    
    return (mean, variance)

In [10]:
mean, variance = fdist_mean_var(5, 10)
print("Mean:", mean)
print("Variance:", variance)

Mean: 1.25
Variance: 1.3541666666666667


------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

Ans. 

Null Hypothesis(H0):- s1 = s2

Alternate Hypothesis(H1):- s1 != s2

Performing F-test we get F-Value to be:-

F-val = s1/s2

=> 1.25

calculating degree of freedom:

dof1 = 10 - 1 => 9

dof2 = 15 - 1 => 14

For significance level of 10% and dof1 = 9 and dof2 = 14 we get critical value equal to:-

critical value = 2.12

Coclusion:-

Since, F-val(1.25) < Critical val(2.12) we fail to reject Null Hypothesis and we can say that we don't have enough evidence to say that there is significant difference between variances
at 10% significance level.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday
night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5%
significance level to determine if the variances are significantly different.


Ans.

Null Hypothesis(H0): varA = varB

Alternate Hypothesis(H1): varA != varB

A = (24, 25, 28, 23, 22, 20, 27)

B = (31, 33, 35, 30, 32, 36)

We Know variance = summation(xi - xbar)^2/n-1

For A xbar = 24.14

For B xbar = 32.83

Using formula we get variances as:-


A = [24, 25, 28, 23, 22, 20, 27]
B = [31, 33, 35, 30, 32, 36]

import numpy as np

varA = np.var(A, ddof =1)
varB = np.var(B, ddof =1)
print(varA)
print(varB)

Performing F-statistics we get:-

F-val = varA/varB 

=> 7.809/5.366

=> 1.455

Degree of freedom is:-

dof1 = 7-1

dof1 => 6

dof2 = 6-1

dof2 => 5

For significance level of 5%  and dof1 = 6, dof2 = 5 we get critical value as:-

critical value = 4.95

Conclusion:-

Since, F-value(1.455) < Critical Val(4.95) we fail to reject Null Hypothesis and can say that there is no siginificance difference between variances at significance level of 5%.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83;
Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances
are significantly different.

Ans. 

Null Hypothesis(H0): varA = varB

Alternate Hypothesis(H1): varA != varB

A = (80, 85, 90, 92, 87, 83)

B = (75, 78, 82, 79, 81, 84)

We Know variance = summation(xi - xbar)^2/n-1

For A xbar = 86.1667

For B xbar = 79.8333

Using formula we get variances as:-

In [14]:
import numpy as np

A = [80, 85, 90, 92, 87, 83]
B = [75, 78, 82, 79, 81, 84]

varA = np.var(A)
varB = np.var(B)
print(varA,varB)

16.472222222222225 8.472222222222221


Performing F-statistics we get:-

F-val = varA/varB 

=> 16.472/8.472

=> 1.944

Degree of freedom is:-

dof1 = 6-1

dof1 => 5

dof2 = 6-1

dof2 => 5

For significance level of 1%  and dof1 = 5, dof2 = 5 we get critical value as:-

critical value = 10.967

Conclusion:-

Since, F-value(1.944) < Critical Val(10.967) we fail to reject Null Hypothesis and can say that there is no siginificance difference between variances at significance level of 1%.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------